File: src/org/lineageos/xiaomi_bluetooth/settings/configs/ListController.java
package org.lineageos.xiaomi_bluetooth.settings.configs;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.preference.ListPreference;
import androidx.preference.Preference;

import org.lineageos.xiaomi_bluetooth.mma.MMADevice;
import org.lineageos.xiaomi_bluetooth.utils.CommonUtils;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;


public abstract class ListController extends ConfigController {

    private static final String TAG = ListController.class.getSimpleName();
    private static final boolean DEBUG = true;

    private boolean isModesUpdated = false;

    public ListController(@NonNull Context context, @NonNull String preferenceKey) {
        super(context, preferenceKey);
    }

    @NonNull
    protected abstract List<ConfigState> getConfigStates();

    @NonNull
    protected abstract ConfigState getDefaultState();

    @Override
    public void displayPreference(@NonNull Preference preference) {
        preference.setDefaultValue(CommonUtils.bytesToHex(getDefaultState().configValue));
        super.displayPreference(preference);
    }

    @Override
    public void updateState(@NonNull Preference preference) {
        if (!(preference instanceof ListPreference listPreference)) {
            Log.w(TAG, "updateState: Incorrect preference type for " + preferenceKey);
            return;
        }

        updateSupportedModes(listPreference);
        super.updateState(preference);
    }

    @Override
    public Available isAvailable() {
        return switch (super.isAvailable()) {
            case AVAILABLE -> !getConfigStates().isEmpty()
                    ? Available.AVAILABLE
                    : Available.UNAVAILABLE;
            case UNAVAILABLE -> Available.UNAVAILABLE;
            case UNKNOWN -> Available.UNKNOWN;
        };
    }

    @Override
    protected void updateValue(@NonNull Preference preference) {
        if (getConfigValue() == null) return;
        if (preference instanceof ListPreference listPreference) {
            listPreference.setValue(CommonUtils.bytesToHex(getConfigValue()));
        }
    }

    private void updateSupportedModes(@NonNull ListPreference preference) {
        if (isAvailable() != Available.AVAILABLE || isModesUpdated) {
            return;
        }
        isModesUpdated = true;

        List<ConfigState> configStates = getConfigStates();
        String[] entryValues = configStates.stream()
                .map(state -> CommonUtils.bytesToHex(state.configValue))
                .toArray(String[]::new);
        String[] entries = configStates.stream()
                .map(state -> context.getString(state.summaryResId))
                .toArray(String[]::new);

        if (DEBUG) {
            Log.d(TAG, String.format("updateSupportedModes: entryValues=%s, entries=%s",
                    Arrays.toString(entryValues), Arrays.toString(entries)));
        }

        preference.setEntryValues(entryValues);
        preference.setEntries(entries);
    }

    @Override
    public String getSummary() {
        if (getConfigValue() == null) {
            return null;
        }

        int resId = getConfigStates().stream()
                .filter(state -> Arrays.equals(state.configValue, getConfigValue()))
                .findFirst().orElse(getDefaultState())
                .summaryResId;
        return context.getString(resId);
    }

    @Override
    public boolean saveConfig(@NonNull MMADevice device, @NonNull Object value) throws IOException {
        if (value instanceof String strValue) {
            value = CommonUtils.hexToBytes(strValue);
        }

        return super.saveConfig(device, value);
    }

}

File: src/org/lineageos/xiaomi_bluetooth/configs/ListController.kt
package org.lineageos.xiaomi_bluetooth.configs

import android.Manifest
import android.content.Context
import android.util.Log
import androidx.annotation.RequiresPermission
import androidx.preference.ListPreference
import androidx.preference.Preference
import org.lineageos.xiaomi_bluetooth.mma.MMADevice
import org.lineageos.xiaomi_bluetooth.utils.ByteUtils.hexToBytes
import org.lineageos.xiaomi_bluetooth.utils.ByteUtils.toHexString

abstract class ListController(context: Context, preferenceKey: String) :
    ConfigController(context, preferenceKey) {

    private var isModesUpdated = false

    protected abstract val configStates: Set<ConfigState>
    protected abstract val defaultState: ConfigState

    override fun updateState(preference: Preference) {
        updateSupportedModes(preference as ListPreference)
        super.updateState(preference)
    }

    override val isAvailable: Available
        get() = when (super.isAvailable) {
            Available.AVAILABLE -> if (configStates.isNotEmpty()) {
                Available.AVAILABLE
            } else {
                Available.UNAVAILABLE
            }

            Available.UNAVAILABLE -> Available.UNAVAILABLE
            Available.UNKNOWN -> Available.UNKNOWN
        }

    override val summary: String?
        get() = configValue?.let {
            context.getString(
                configStates.first { state -> state.configValue.contentEquals(configValue) }.summaryResId
            )
        }

    override fun updateValue(preference: Preference) {
        configValue?.run {
            (preference as ListPreference).value = toHexString()
        }
    }

    private fun updateSupportedModes(preference: ListPreference) {
        if (isAvailable != Available.AVAILABLE || isModesUpdated) {
            return
        }
        isModesUpdated = true

        val entryValues = configStates
            .map { state -> state.configValue.toHexString() }
            .toTypedArray()
        val entries = configStates
            .map { state -> context.getString(state.summaryResId) }
            .toTypedArray()

        if (DEBUG) Log.d(
            TAG,
            "updateSupportedModes: " +
                    "entryValues=${entryValues.contentToString()}, " +
                    "entries=${entries.contentToString()}"
        )

        preference.entryValues = entryValues
        preference.entries = entries
    }

    @RequiresPermission(allOf = [Manifest.permission.BLUETOOTH_SCAN, Manifest.permission.BLUETOOTH_CONNECT])
    override fun saveConfig(device: MMADevice, value: Any): Boolean {
        return super.saveConfig(device, if (value is String) value.hexToBytes() else value)
    }

    companion object {
        private val TAG = ListController::class.java.simpleName
        private const val DEBUG = true
    }
}
